src/nlp/utils.jl

```diff
@@ -131,5 +131,7 @@ macro default_counters(Model, inner)
   end))
   push!(ex.args, :(NLPModels.increment!(nlp::$(esc(Model)), s::Symbol) = increment!(nlp.$inner, s)))
   push!(ex.args, :(NLPModels.decrement!(nlp::$(esc(Model)), s::Symbol) = decrement!(nlp.$inner, s)))
+
+  push!(ex.args, :(Base.getproperty(nlp::$(esc(Model)), s::Symbol) = ( s == :counters ? nlp.$inner.counters : getfield(nlp ,s))))
```

Comment:
General question, is there a difference between:
- nlp.$inner.counters
- getfield(nlp.$inner, :counters)
- getfield(getfield(nlp, $inner), :counters)

in particular if you check the application of this:
```@code_warntype quasinewton_model.counters```

Reply:
I think there is no difference but I am not sure

Reply:
In the first case,
```
julia> @code_warntype nlp.counters
MethodInstance for getproperty(::NLPModelsModifiers.DiagonalQNModel{Float64, Vector{Float64}, ADNLPModel{Float64, Vector{Float64}, Vector{Int64}}, NLPModelMeta{Float64, Vector{Float64}}, LinearOperators.SpectralGradient{Float64, Int64, LinearOperators.var"#100#101"{Vector{Float64}}}}, ::Symbol)
  from getproperty(nlp::QuasiNewtonModel, s::Symbol) @ NLPModelsModifiers C:\Users\mgoll\NLPModels.jl\src\nlp\utils.jl:135
Arguments
  #self#::Core.Const(getproperty)
  nlp::NLPModelsModifiers.DiagonalQNModel{Float64, Vector{Float64}, ADNLPModel{Float64, Vector{Float64}, Vector{Int64}}, NLPModelMeta{Float64, Vector{Float64}}, LinearOperators.SpectralGradient{Float64, Int64, LinearOperators.var"#100#101"{Vector{Float64}}}}
  s::Symbol
Body::Any
1 ─ %1 = (s == :counters)::Bool
└──      goto #3 if not %1
2 ─ %3 = Base.getproperty(nlp, :model)::ADNLPModel{Float64, Vector{Float64}, Vector{Int64}}
│   %4 = Base.getproperty(%3, :counters)::Counters
└──      return %4
3 ─ %6 = NLPModels.getfield(nlp, s)::Union{ADNLPModel{Float64, Vector{Float64}, Vector{Int64}}, LinearOperators.SpectralGradient{Float64, Int64, LinearOperators.var"#100#101"{Vector{Float64}}}, NLPModelMeta{Float64, Vector{Float64}}}
└──      return %6
```
In the second case, 
```
julia> @code_warntype nlp.counters
MethodInstance for getproperty(::NLPModelsModifiers.DiagonalQNModel{Float64, Vector{Float64}, ADNLPModel{Float64, Vector{Float64}, Vector{Int64}}, NLPModelMeta{Float64, Vector{Float64}}, LinearOperators.SpectralGradient{Float64, Int64, LinearOperators.var"#100#101"{Vector{Float64}}}}, ::Symbol)
  from getproperty(nlp::QuasiNewtonModel, s::Symbol) @ NLPModelsModifiers C:\Users\mgoll\NLPModels.jl\src\nlp\utils.jl:135
Arguments
  #self#::Core.Const(getproperty)
  nlp::NLPModelsModifiers.DiagonalQNModel{Float64, Vector{Float64}, ADNLPModel{Float64, Vector{Float64}, Vector{Int64}}, NLPModelMeta{Float64, Vector{Float64}}, LinearOperators.SpectralGradient{Float64, Int64, LinearOperators.var"#100#101"{Vector{Float64}}}}
  s::Symbol
Body::Any
1 ─ %1 = (s == :counters)::Bool
└──      goto #3 if not %1
2 ─ %3 = Base.getproperty(nlp, :model)::ADNLPModel{Float64, Vector{Float64}, Vector{Int64}}
│   %4 = NLPModels.getfield(%3, :counters)::Counters
└──      return %4
3 ─ %6 = NLPModels.getfield(nlp, s)::Union{ADNLPModel{Float64, Vector{Float64}, Vector{Int64}}, LinearOperators.SpectralGradient{Float64, Int64, LinearOperators.var"#100#101"{Vector{Float64}}}, NLPModelMeta{Float64, Vector{Float64}}}
└──      return %6
```
In the third case,
```
julia> @code_warntype nlp.counters
MethodInstance for getproperty(::NLPModelsModifiers.DiagonalQNModel{Float64, Vector{Float64}, ADNLPModel{Float64, Vector{Float64}, Vector{Int64}}, NLPModelMeta{Float64, Vector{Float64}}, LinearOperators.SpectralGradient{Float64, Int64, LinearOperators.var"#100#101"{Vector{Float64}}}}, ::Symbol)
  from getproperty(nlp::QuasiNewtonModel, s::Symbol) @ NLPModelsModifiers C:\Users\mgoll\NLPModels.jl\src\nlp\utils.jl:135
Arguments
  #self#::Core.Const(getproperty)
  nlp::NLPModelsModifiers.DiagonalQNModel{Float64, Vector{Float64}, ADNLPModel{Float64, Vector{Float64}, Vector{Int64}}, NLPModelMeta{Float64, Vector{Float64}}, LinearOperators.SpectralGradient{Float64, Int64, LinearOperators.var"#100#101"{Vector{Float64}}}}
  s::Symbol
Body::Any
1 ─ %1 = (s == :counters)::Bool
└──      goto #3 if not %1
2 ─ %3 = NLPModels.getfield(nlp, NLPModels.model)::Union{ADNLPModel{Float64, Vector{Float64}, Vector{Int64}}, LinearOperators.SpectralGradient{Float64, Int64, LinearOperators.var"#100#101"{Vector{Float64}}}, NLPModelMeta{Float64, Vector{Float64}}}
│   %4 = NLPModels.getfield(%3, :counters)::Counters
└──      return %4
3 ─ %6 = NLPModels.getfield(nlp, s)::Union{ADNLPModel{Float64, Vector{Float64}, Vector{Int64}}, LinearOperators.SpectralGradient{Float64, Int64, LinearOperators.var"#100#101"{Vector{Float64}}}, NLPModelMeta{Float64, Vector{Float64}}}
└──      return %6
```

Reply:
Thanks @MaxenceGollier